spacy/kb/candidate.py

```diff
@@ -0,0 +1,118 @@
+import abc
+from typing import List, Union, Callable
+
+
+class Candidate(abc.ABC):
+    """A `Candidate` object refers to a textual mention (`alias`) that may or may not be resolved
+    to a specific `entity_id` from a Knowledge Base. This will be used as input for the entity_id linking
+    algorithm which will disambiguate the various candidates to the correct one.
+    Each candidate (alias, entity_id) pair is assigned a certain prior probability.
+
+    DOCS: https://spacy.io/api/kb/#candidate-init
+    """
+
+    def __init__(
+        self,
+        mention: str,
+        entity_id: int,
+        entity_name: str,
+        entity_vector: List[float],
+        prior_prob: float,
+    ):
+        """Initializes properties of `Candidate` instance.
+        mention (str): Mention text for this candidate.
+        entity_id (int): Unique entity ID.
+        entity_name (str): Entity name.
+        entity_vector (List[float]): Entity embedding.
+        prior_prob (float): Prior probability of entity for this mention - i.e. the probability that, independent of
+            the context, this mention resolves to this entity_id in the corpus used to build the knowledge base. In
+            cases in which this isn't always possible (e.g.: the corpus to analyse contains mentions that the KB corpus
+            doesn't) it might be better to eschew this information and always supply the same value.
+        """
+        self._mention = mention
+        self._entity_id = entity_id
+        self._entity_name = entity_name
+        self._entity_vector = entity_vector
+        self._prior_prob = prior_prob
+
+    @property
+    def entity(self) -> int:
+        """RETURNS (int): Unique entity ID."""
+        return self._entity_id
+
+    @property
+    def entity_(self) -> str:
+        """RETURNS (int): Entity name."""
+        return self._entity_name
```

Comment:
I know how this historically came to be because of storing the hashes etc, but I want to take a moment and consider whether we still want this. Also whether we should insist that the `entity_id` is an `int`, which might get annoying for use-cases where the KB identifiers are actually strings, like for Wikidata.

Reply:
I actually had this as `Union[str, int]` recently. I reverted to `int` because `v4` [changes](https://github.com/explosion/spaCy/blob/v4/spacy/pipeline/entity_linker.py#L588) expect `.entity` to be an `int` and I'm not sure how trivial it would be to modify the `_add_activation` apparatus.

That said, we could always allow passing `entity_id` as `Union[int, str]` - if it's a `str`, hash it in `Candidate` and return the internal hash `int` in `.entity`, if it's an `int`, return that. Does that make sense to you?

Reply:
I can just draft this, if it's easier to discuss then.

Reply:
I knew I'd seen `Union[str, int]` somewhere! Yea, I think that's better. And then we can have a property `entity_hash` that returns an `int` by hashing if it's a `str`. So yea, I think we're proposing the same thing.

Reply:
I restructured `Candidate` attributes in https://github.com/explosion/spaCy/pull/12341/commits/38dce966e55dbeb61d1d085ff97ce50f5095dea8. This includes doing away with storing the mention hash, since in the current version `Candidate` stores the mention string anyway; renaming attributes for more clarity; and using `Union[str, int]` for `entity_id`, with support for converting `entity_id` into a string or integer respectively.

Consider this a draft and let me know what you think.